# Action units - cleaning

Output
- /Users/dionnespaltman/Desktop/V3/action_units_temp.pkl
- /Users/dionnespaltman/Desktop/V3/action_units_temp.csv

In [18]:
# import 
import zipfile
import os
import pandas as pd
import csv
import socket  
import pickle
import re

## Unzipping

In [ ]:
# Define paths
zip_file_path = '/Users/dionnespaltman/Desktop/downloading/full donation data.zip'
output_folder = '/Users/dionnespaltman/Desktop/downloading/full donation data - unzipped'

# Create the output folder if it doesn't exist
if not os.path.exists(output_folder):
    os.makedirs(output_folder)

# Unzip the folder
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(output_folder)

print("Unzipped folder created at:", output_folder)

In [3]:
# Define the path to the folder
folder_path = '/Users/dionnespaltman/Desktop/downloading/full donation data - unzipped/full donation data'

# Initialize total size variable
total_size = 0

# Iterate over all files in the folder
for filename in os.listdir(folder_path):
    file_path = os.path.join(folder_path, filename)
    # Check if it's a file (not a directory)
    if os.path.isfile(file_path):
        # Get the size of the file and add it to total_size
        total_size += os.path.getsize(file_path)

# Convert total size to a human-readable format (e.g., bytes to megabytes)
total_size_mb = total_size / (1024 * 1024)  # Convert bytes to megabytes

print("Total size of all files in the folder:", total_size_mb, "MB")

Total size of all files in the folder: 18684.90839099884 MB


## Creating a dictionary with all the files 

This takes around 8-14 minutes. 

In [3]:
# Define the path to the folder containing the unzipped files
folder_path = '/Users/dionnespaltman/Desktop/downloading/full donation data - unzipped/full donation data'

# List all files in the folder
file_names = os.listdir(folder_path)

# Read a subset of files into a dictionary
data = {}
num_files_to_read = 450  # Adjust the number of files to read as needed
for i, file_name in enumerate(file_names):
    if i >= num_files_to_read:
        break
    if file_name.endswith('.csv'):  # Assuming the files are CSV format
        file_path = os.path.join(folder_path, file_name)
        try:
            data[file_name] = pd.read_csv(file_path)
            print("File loaded successfully:", file_name)
        except Exception as e:
            print("Error loading file:", file_name, "- Error:", e)

# Now you have a dictionary 'data' containing DataFrames for each file (up to the specified number)
# Access them using keys (file names)
# For example:
print("Number of files loaded:", len(data))

File loaded successfully: 16-07.csv
File loaded successfully: 80_04,05,06.csv
File loaded successfully: 7-04-05-06.csv
File loaded successfully: 324_02.csv
File loaded successfully: 87_04,05,06.csv
File loaded successfully: 78_04,05,06.csv
File loaded successfully: 328_02.csv
File loaded successfully: 105_04 donation not completed. No blood flow.csv
File loaded successfully: 92_04,05,06.csv
File loaded successfully: 38_4,5,6.csv
File loaded successfully: 129_04,05,06.csv
File loaded successfully: 85_07.csv
File loaded successfully: DSCN2370.csv
File loaded successfully: 95_04,05,06.csv
File loaded successfully: 300_02.csv
File loaded successfully: 20-07.csv
File loaded successfully: 97_07.csv
File loaded successfully: 290_01.csv
File loaded successfully: 312_02.csv
File loaded successfully: 127_03.csv
File loaded successfully: 40_07.csv
File loaded successfully: 101_07.csv
File loaded successfully: 118_04,05,06.csv
File loaded successfully: 14-03,04,05,06.csv
File loaded successfully: 

/var/folders/_4/nzq6mygj7j71_l3z_c9kc7wr0000gn/T/ipykernel_7071/3814314940.py:16: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  data[file_name] = pd.read_csv(file_path)


File loaded successfully: 108_04,05,06.csv
File loaded successfully: 80_07.csv
File loaded successfully: 305_02.csv
File loaded successfully: 299_01.csv
Number of files loaded: 403


In [4]:
# Check if the dictionary is not empty
if data:
    # Get the first key (file name) and its corresponding DataFrame
    first_file_name = next(iter(data.keys()))
    first_df = data[first_file_name]
    
    # Display the DataFrame
    print("DataFrame for the first file '{}' in the dictionary:".format(first_file_name))
    display(first_df.head(10))
else:
    print("The dictionary is empty. No files loaded.")

DataFrame for the first file '16-07.csv' in the dictionary:


frame   face_id   timestamp   confidence   success   gaze_0_x   gaze_0_y  \
0      1         0        0.00         0.98         1  -0.186231  -0.071084   
1      2         0        0.04         0.98         1  -0.195272  -0.066113   
2      3         0        0.08         0.98         1  -0.183891  -0.072790   
3      4         0        0.12         0.98         1  -0.189050  -0.061480   
4      5         0        0.16         0.98         1  -0.193372  -0.059505   
5      6         0        0.20         0.98         1  -0.182038  -0.053976   
6      7         0        0.24         0.98         1  -0.174060  -0.062096   
7      8         0        0.28         0.98         1  -0.183968  -0.068336   
8      9         0        0.32         0.98         1  -0.172816  -0.076083   
9     10         0        0.36         0.98         1  -0.162287  -0.070443   

    gaze_0_z   gaze_1_x   gaze_1_y  ...   AU12_c   AU14_c   AU15_c   AU17_c  \
0  -0.979931  -0.321020  -0.025671  ...      0.0      1.0      0.0      0.0   
1  -0.978518  -0.321741  -0.022495  ...      1.0      1.0      0.0      0.0   
2  -0.980248  -0.325824  -0.032522  ...      0.0      1.0      0.0      1.0   
3  -0.980041  -0.323833  -0.018319  ...      0.0      1.0      0.0      0.0   
4  -0.979319  -0.317909  -0.020461  ...      0.0      1.0      0.0      0.0   
5  -0.981809  -0.320776  -0.025344  ...      0.0      1.0      0.0      0.0   
6  -0.982775  -0.332763  -0.027208  ...      0.0      1.0      0.0      0.0   
7  -0.980554  -0.325630  -0.040619  ...      0.0      1.0      0.0      0.0   
8  -0.982011  -0.330309  -0.047691  ...      0.0      1.0      0.0      0.0   
9  -0.984226  -0.325321  -0.033416  ...      0.0      1.0      0.0      0.0   

    AU20_c   AU23_c   AU25_c   AU26_c   AU28_c   AU45_c  
0      0.0      1.0      0.0      0.0      0.0      0.0  
1      0.0      1.0      0.0      0.0      0.0      0.0  
2      0.0      1.0      0.0      0.0      0.0      0.0  
3      0.0      1.0      0.0      0.0      0.0      0.0  
4      0.0      1.0      0.0      0.0      0.0      0.0  
5      0.0      1.0      0.0      0.0      0.0      0.0  
6      0.0      1.0      0.0      0.0      0.0      0.0  
7      0.0      1.0      0.0      0.0      0.0      0.0  
8      0.0      1.0      0.0      0.0      0.0      0.0  
9      0.0      1.0      0.0      0.0      0.0      0.0  

[10 rows x 714 columns]

Saving the dictionary with all files. 

In [5]:
# Specify the file path to save the dictionary
save_path = '/Users/dionnespaltman/Desktop/V3/data_dictionary_temp.pkl'

# Save the dictionary to a file using pickle
with open(save_path, 'wb') as file:
    pickle.dump(data, file)

print("Dictionary saved to:", save_path)

Dictionary saved to: /Users/dionnespaltman/Desktop/V3/data_dictionary_temp.pkl


## Loading the dictionary 

In [ ]:
# Specify the file path from which to load the dictionary
load_path = '/Users/dionnespaltman/Desktop/V3/data_dictionary_temp.pkl'  # Update with the actual file path

# Load the dictionary from the file using pickle
with open(load_path, 'rb') as file:
    data = pickle.load(file)

# print("Dictionary loaded from:", data)

## Dropping columns 

In [6]:
# Check if the dictionary is not empty
if data:
    # Get the first key (file name) and its corresponding DataFrame
    first_file_name = next(iter(data.keys()))
    first_df = data[first_file_name]
    
    # Print the column names of the DataFrame
    print("Column names of the first file '{}' in the dictionary:".format(first_file_name))
    print(first_df.columns.tolist())
else:
    print("The dictionary is empty. No files loaded.")

Column names of the first file '16-07.csv' in the dictionary:
['frame', ' face_id', ' timestamp', ' confidence', ' success', ' gaze_0_x', ' gaze_0_y', ' gaze_0_z', ' gaze_1_x', ' gaze_1_y', ' gaze_1_z', ' gaze_angle_x', ' gaze_angle_y', ' eye_lmk_x_0', ' eye_lmk_x_1', ' eye_lmk_x_2', ' eye_lmk_x_3', ' eye_lmk_x_4', ' eye_lmk_x_5', ' eye_lmk_x_6', ' eye_lmk_x_7', ' eye_lmk_x_8', ' eye_lmk_x_9', ' eye_lmk_x_10', ' eye_lmk_x_11', ' eye_lmk_x_12', ' eye_lmk_x_13', ' eye_lmk_x_14', ' eye_lmk_x_15', ' eye_lmk_x_16', ' eye_lmk_x_17', ' eye_lmk_x_18', ' eye_lmk_x_19', ' eye_lmk_x_20', ' eye_lmk_x_21', ' eye_lmk_x_22', ' eye_lmk_x_23', ' eye_lmk_x_24', ' eye_lmk_x_25', ' eye_lmk_x_26', ' eye_lmk_x_27', ' eye_lmk_x_28', ' eye_lmk_x_29', ' eye_lmk_x_30', ' eye_lmk_x_31', ' eye_lmk_x_32', ' eye_lmk_x_33', ' eye_lmk_x_34', ' eye_lmk_x_35', ' eye_lmk_x_36', ' eye_lmk_x_37', ' eye_lmk_x_38', ' eye_lmk_x_39', ' eye_lmk_x_40', ' eye_lmk_x_41', ' eye_lmk_x_42', ' eye_lmk_x_43', ' eye_lmk_x_44', ' eye_lm

The column names: ['frame', ' face_id', ' timestamp', ' confidence', ' success', ' gaze_0_x', ' gaze_0_y', ' gaze_0_z', ' gaze_1_x', ' gaze_1_y', ' gaze_1_z', ' gaze_angle_x', ' gaze_angle_y', ' eye_lmk_x_0', ' eye_lmk_x_1', ' eye_lmk_x_2', ' eye_lmk_x_3', ' eye_lmk_x_4', ' eye_lmk_x_5', ' eye_lmk_x_6', ' eye_lmk_x_7', ' eye_lmk_x_8', ' eye_lmk_x_9', ' eye_lmk_x_10', ' eye_lmk_x_11', ' eye_lmk_x_12', ' eye_lmk_x_13', ' eye_lmk_x_14', ' eye_lmk_x_15', ' eye_lmk_x_16', ' eye_lmk_x_17', ' eye_lmk_x_18', ' eye_lmk_x_19', ' eye_lmk_x_20', ' eye_lmk_x_21', ' eye_lmk_x_22', ' eye_lmk_x_23', ' eye_lmk_x_24', ' eye_lmk_x_25', ' eye_lmk_x_26', ' eye_lmk_x_27', ' eye_lmk_x_28', ' eye_lmk_x_29', ' eye_lmk_x_30', ' eye_lmk_x_31', ' eye_lmk_x_32', ' eye_lmk_x_33', ' eye_lmk_x_34', ' eye_lmk_x_35', ' eye_lmk_x_36', ' eye_lmk_x_37', ' eye_lmk_x_38', ' eye_lmk_x_39', ' eye_lmk_x_40', ' eye_lmk_x_41', ' eye_lmk_x_42', ' eye_lmk_x_43', ' eye_lmk_x_44', ' eye_lmk_x_45', ' eye_lmk_x_46', ' eye_lmk_x_47', ' eye_lmk_x_48', ' eye_lmk_x_49', ' eye_lmk_x_50', ' eye_lmk_x_51', ' eye_lmk_x_52', ' eye_lmk_x_53', ' eye_lmk_x_54', ' eye_lmk_x_55', ' eye_lmk_y_0', ' eye_lmk_y_1', ' eye_lmk_y_2', ' eye_lmk_y_3', ' eye_lmk_y_4', ' eye_lmk_y_5', ' eye_lmk_y_6', ' eye_lmk_y_7', ' eye_lmk_y_8', ' eye_lmk_y_9', ' eye_lmk_y_10', ' eye_lmk_y_11', ' eye_lmk_y_12', ' eye_lmk_y_13', ' eye_lmk_y_14', ' eye_lmk_y_15', ' eye_lmk_y_16', ' eye_lmk_y_17', ' eye_lmk_y_18', ' eye_lmk_y_19', ' eye_lmk_y_20', ' eye_lmk_y_21', ' eye_lmk_y_22', ' eye_lmk_y_23', ' eye_lmk_y_24', ' eye_lmk_y_25', ' eye_lmk_y_26', ' eye_lmk_y_27', ' eye_lmk_y_28', ' eye_lmk_y_29', ' eye_lmk_y_30', ' eye_lmk_y_31', ' eye_lmk_y_32', ' eye_lmk_y_33', ' eye_lmk_y_34', ' eye_lmk_y_35', ' eye_lmk_y_36', ' eye_lmk_y_37', ' eye_lmk_y_38', ' eye_lmk_y_39', ' eye_lmk_y_40', ' eye_lmk_y_41', ' eye_lmk_y_42', ' eye_lmk_y_43', ' eye_lmk_y_44', ' eye_lmk_y_45', ' eye_lmk_y_46', ' eye_lmk_y_47', ' eye_lmk_y_48', ' eye_lmk_y_49', ' eye_lmk_y_50', ' eye_lmk_y_51', ' eye_lmk_y_52', ' eye_lmk_y_53', ' eye_lmk_y_54', ' eye_lmk_y_55', ' eye_lmk_X_0', ' eye_lmk_X_1', ' eye_lmk_X_2', ' eye_lmk_X_3', ' eye_lmk_X_4', ' eye_lmk_X_5', ' eye_lmk_X_6', ' eye_lmk_X_7', ' eye_lmk_X_8', ' eye_lmk_X_9', ' eye_lmk_X_10', ' eye_lmk_X_11', ' eye_lmk_X_12', ' eye_lmk_X_13', ' eye_lmk_X_14', ' eye_lmk_X_15', ' eye_lmk_X_16', ' eye_lmk_X_17', ' eye_lmk_X_18', ' eye_lmk_X_19', ' eye_lmk_X_20', ' eye_lmk_X_21', ' eye_lmk_X_22', ' eye_lmk_X_23', ' eye_lmk_X_24', ' eye_lmk_X_25', ' eye_lmk_X_26', ' eye_lmk_X_27', ' eye_lmk_X_28', ' eye_lmk_X_29', ' eye_lmk_X_30', ' eye_lmk_X_31', ' eye_lmk_X_32', ' eye_lmk_X_33', ' eye_lmk_X_34', ' eye_lmk_X_35', ' eye_lmk_X_36', ' eye_lmk_X_37', ' eye_lmk_X_38', ' eye_lmk_X_39', ' eye_lmk_X_40', ' eye_lmk_X_41', ' eye_lmk_X_42', ' eye_lmk_X_43', ' eye_lmk_X_44', ' eye_lmk_X_45', ' eye_lmk_X_46', ' eye_lmk_X_47', ' eye_lmk_X_48', ' eye_lmk_X_49', ' eye_lmk_X_50', ' eye_lmk_X_51', ' eye_lmk_X_52', ' eye_lmk_X_53', ' eye_lmk_X_54', ' eye_lmk_X_55', ' eye_lmk_Y_0', ' eye_lmk_Y_1', ' eye_lmk_Y_2', ' eye_lmk_Y_3', ' eye_lmk_Y_4', ' eye_lmk_Y_5', ' eye_lmk_Y_6', ' eye_lmk_Y_7', ' eye_lmk_Y_8', ' eye_lmk_Y_9', ' eye_lmk_Y_10', ' eye_lmk_Y_11', ' eye_lmk_Y_12', ' eye_lmk_Y_13', ' eye_lmk_Y_14', ' eye_lmk_Y_15', ' eye_lmk_Y_16', ' eye_lmk_Y_17', ' eye_lmk_Y_18', ' eye_lmk_Y_19', ' eye_lmk_Y_20', ' eye_lmk_Y_21', ' eye_lmk_Y_22', ' eye_lmk_Y_23', ' eye_lmk_Y_24', ' eye_lmk_Y_25', ' eye_lmk_Y_26', ' eye_lmk_Y_27', ' eye_lmk_Y_28', ' eye_lmk_Y_29', ' eye_lmk_Y_30', ' eye_lmk_Y_31', ' eye_lmk_Y_32', ' eye_lmk_Y_33', ' eye_lmk_Y_34', ' eye_lmk_Y_35', ' eye_lmk_Y_36', ' eye_lmk_Y_37', ' eye_lmk_Y_38', ' eye_lmk_Y_39', ' eye_lmk_Y_40', ' eye_lmk_Y_41', ' eye_lmk_Y_42', ' eye_lmk_Y_43', ' eye_lmk_Y_44', ' eye_lmk_Y_45', ' eye_lmk_Y_46', ' eye_lmk_Y_47', ' eye_lmk_Y_48', ' eye_lmk_Y_49', ' eye_lmk_Y_50', ' eye_lmk_Y_51', ' eye_lmk_Y_52', ' eye_lmk_Y_53', ' eye_lmk_Y_54', ' eye_lmk_Y_55', ' eye_lmk_Z_0', ' eye_lmk_Z_1', ' eye_lmk_Z_2', ' eye_lmk_Z_3', ' eye_lmk_Z_4', ' eye_lmk_Z_5', ' eye_lmk_Z_6', ' eye_lmk_Z_7', ' eye_lmk_Z_8', ' eye_lmk_Z_9', ' eye_lmk_Z_10', ' eye_lmk_Z_11', ' eye_lmk_Z_12', ' eye_lmk_Z_13', ' eye_lmk_Z_14', ' eye_lmk_Z_15', ' eye_lmk_Z_16', ' eye_lmk_Z_17', ' eye_lmk_Z_18', ' eye_lmk_Z_19', ' eye_lmk_Z_20', ' eye_lmk_Z_21', ' eye_lmk_Z_22', ' eye_lmk_Z_23', ' eye_lmk_Z_24', ' eye_lmk_Z_25', ' eye_lmk_Z_26', ' eye_lmk_Z_27', ' eye_lmk_Z_28', ' eye_lmk_Z_29', ' eye_lmk_Z_30', ' eye_lmk_Z_31', ' eye_lmk_Z_32', ' eye_lmk_Z_33', ' eye_lmk_Z_34', ' eye_lmk_Z_35', ' eye_lmk_Z_36', ' eye_lmk_Z_37', ' eye_lmk_Z_38', ' eye_lmk_Z_39', ' eye_lmk_Z_40', ' eye_lmk_Z_41', ' eye_lmk_Z_42', ' eye_lmk_Z_43', ' eye_lmk_Z_44', ' eye_lmk_Z_45', ' eye_lmk_Z_46', ' eye_lmk_Z_47', ' eye_lmk_Z_48', ' eye_lmk_Z_49', ' eye_lmk_Z_50', ' eye_lmk_Z_51', ' eye_lmk_Z_52', ' eye_lmk_Z_53', ' eye_lmk_Z_54', ' eye_lmk_Z_55', ' pose_Tx', ' pose_Ty', ' pose_Tz', ' pose_Rx', ' pose_Ry', ' pose_Rz', ' x_0', ' x_1', ' x_2', ' x_3', ' x_4', ' x_5', ' x_6', ' x_7', ' x_8', ' x_9', ' x_10', ' x_11', ' x_12', ' x_13', ' x_14', ' x_15', ' x_16', ' x_17', ' x_18', ' x_19', ' x_20', ' x_21', ' x_22', ' x_23', ' x_24', ' x_25', ' x_26', ' x_27', ' x_28', ' x_29', ' x_30', ' x_31', ' x_32', ' x_33', ' x_34', ' x_35', ' x_36', ' x_37', ' x_38', ' x_39', ' x_40', ' x_41', ' x_42', ' x_43', ' x_44', ' x_45', ' x_46', ' x_47', ' x_48', ' x_49', ' x_50', ' x_51', ' x_52', ' x_53', ' x_54', ' x_55', ' x_56', ' x_57', ' x_58', ' x_59', ' x_60', ' x_61', ' x_62', ' x_63', ' x_64', ' x_65', ' x_66', ' x_67', ' y_0', ' y_1', ' y_2', ' y_3', ' y_4', ' y_5', ' y_6', ' y_7', ' y_8', ' y_9', ' y_10', ' y_11', ' y_12', ' y_13', ' y_14', ' y_15', ' y_16', ' y_17', ' y_18', ' y_19', ' y_20', ' y_21', ' y_22', ' y_23', ' y_24', ' y_25', ' y_26', ' y_27', ' y_28', ' y_29', ' y_30', ' y_31', ' y_32', ' y_33', ' y_34', ' y_35', ' y_36', ' y_37', ' y_38', ' y_39', ' y_40', ' y_41', ' y_42', ' y_43', ' y_44', ' y_45', ' y_46', ' y_47', ' y_48', ' y_49', ' y_50', ' y_51', ' y_52', ' y_53', ' y_54', ' y_55', ' y_56', ' y_57', ' y_58', ' y_59', ' y_60', ' y_61', ' y_62', ' y_63', ' y_64', ' y_65', ' y_66', ' y_67', ' X_0', ' X_1', ' X_2', ' X_3', ' X_4', ' X_5', ' X_6', ' X_7', ' X_8', ' X_9', ' X_10', ' X_11', ' X_12', ' X_13', ' X_14', ' X_15', ' X_16', ' X_17', ' X_18', ' X_19', ' X_20', ' X_21', ' X_22', ' X_23', ' X_24', ' X_25', ' X_26', ' X_27', ' X_28', ' X_29', ' X_30', ' X_31', ' X_32', ' X_33', ' X_34', ' X_35', ' X_36', ' X_37', ' X_38', ' X_39', ' X_40', ' X_41', ' X_42', ' X_43', ' X_44', ' X_45', ' X_46', ' X_47', ' X_48', ' X_49', ' X_50', ' X_51', ' X_52', ' X_53', ' X_54', ' X_55', ' X_56', ' X_57', ' X_58', ' X_59', ' X_60', ' X_61', ' X_62', ' X_63', ' X_64', ' X_65', ' X_66', ' X_67', ' Y_0', ' Y_1', ' Y_2', ' Y_3', ' Y_4', ' Y_5', ' Y_6', ' Y_7', ' Y_8', ' Y_9', ' Y_10', ' Y_11', ' Y_12', ' Y_13', ' Y_14', ' Y_15', ' Y_16', ' Y_17', ' Y_18', ' Y_19', ' Y_20', ' Y_21', ' Y_22', ' Y_23', ' Y_24', ' Y_25', ' Y_26', ' Y_27', ' Y_28', ' Y_29', ' Y_30', ' Y_31', ' Y_32', ' Y_33', ' Y_34', ' Y_35', ' Y_36', ' Y_37', ' Y_38', ' Y_39', ' Y_40', ' Y_41', ' Y_42', ' Y_43', ' Y_44', ' Y_45', ' Y_46', ' Y_47', ' Y_48', ' Y_49', ' Y_50', ' Y_51', ' Y_52', ' Y_53', ' Y_54', ' Y_55', ' Y_56', ' Y_57', ' Y_58', ' Y_59', ' Y_60', ' Y_61', ' Y_62', ' Y_63', ' Y_64', ' Y_65', ' Y_66', ' Y_67', ' Z_0', ' Z_1', ' Z_2', ' Z_3', ' Z_4', ' Z_5', ' Z_6', ' Z_7', ' Z_8', ' Z_9', ' Z_10', ' Z_11', ' Z_12', ' Z_13', ' Z_14', ' Z_15', ' Z_16', ' Z_17', ' Z_18', ' Z_19', ' Z_20', ' Z_21', ' Z_22', ' Z_23', ' Z_24', ' Z_25', ' Z_26', ' Z_27', ' Z_28', ' Z_29', ' Z_30', ' Z_31', ' Z_32', ' Z_33', ' Z_34', ' Z_35', ' Z_36', ' Z_37', ' Z_38', ' Z_39', ' Z_40', ' Z_41', ' Z_42', ' Z_43', ' Z_44', ' Z_45', ' Z_46', ' Z_47', ' Z_48', ' Z_49', ' Z_50', ' Z_51', ' Z_52', ' Z_53', ' Z_54', ' Z_55', ' Z_56', ' Z_57', ' Z_58', ' Z_59', ' Z_60', ' Z_61', ' Z_62', ' Z_63', ' Z_64', ' Z_65', ' Z_66', ' Z_67', ' p_scale', ' p_rx', ' p_ry', ' p_rz', ' p_tx', ' p_ty', ' p_0', ' p_1', ' p_2', ' p_3', ' p_4', ' p_5', ' p_6', ' p_7', ' p_8', ' p_9', ' p_10', ' p_11', ' p_12', ' p_13', ' p_14', ' p_15', ' p_16', ' p_17', ' p_18', ' p_19', ' p_20', ' p_21', ' p_22', ' p_23', ' p_24', ' p_25', ' p_26', ' p_27', ' p_28', ' p_29', ' p_30', ' p_31', ' p_32', ' p_33', ' AU01_r', ' AU02_r', ' AU04_r', ' AU05_r', ' AU06_r', ' AU07_r', ' AU09_r', ' AU10_r', ' AU12_r', ' AU14_r', ' AU15_r', ' AU17_r', ' AU20_r', ' AU23_r', ' AU25_r', ' AU26_r', ' AU45_r', ' AU01_c', ' AU02_c', ' AU04_c', ' AU05_c', ' AU06_c', ' AU07_c', ' AU09_c', ' AU10_c', ' AU12_c', ' AU14_c', ' AU15_c', ' AU17_c', ' AU20_c', ' AU23_c', ' AU25_c', ' AU26_c', ' AU28_c', ' AU45_c']

In [8]:
# Get the DataFrame with the key '16-07.csv'
df_16_07 = data['16-07.csv']

# Get the shape of the DataFrame
shape_16_07 = df_16_07.shape

# Print the shape
print("Shape of the DataFrame with key '16-07.csv':", shape_16_07)


Shape of the DataFrame with key '16-07.csv': (2750, 714)


In [9]:
def get_columns_to_delete(columns):
    # List of substrings to search for in column names
    substrings = ['gaze', 'p', 'x', 'X', 'Y', 'y', 'Z',  'pose', 'eye']

    # Initialize an empty list to store column names to delete
    columns_to_delete = []

    # Iterate through each column name
    for column in columns:
        # Check if any of the substrings are present in the column name
        if any(sub in column for sub in substrings):
            # If present, add the column name to the list of columns to delete
            columns_to_delete.append(column)

    return columns_to_delete

In [10]:
# Example usage:
columns = ['AU01_r', 'AU02_r', 'AU04_r', 'AU05_r', 'gaze_0_x', 'gaze_0_y', 'pose_Rx', 'pose_Ry']
columns_to_delete = get_columns_to_delete(columns)
print(columns_to_delete)

['gaze_0_x', 'gaze_0_y', 'pose_Rx', 'pose_Ry']


In [11]:
# Get the DataFrame from the dictionary
df_to_process = data['16-07.csv']

# Get the list of column names
columns = df_to_process.columns
print(len(columns))
#print(columns)

# Get the list of columns to delete
columns_to_delete = get_columns_to_delete(columns)

# Print the list of columns to delete
#print(list(columns))
#print(columns_to_delete)
print(len(columns_to_delete))

714
675


In [12]:
# Define the new dictionary to store modified DataFrames
data_dropped = {}

# Iterate through all the DataFrames in the original dictionary
for key, df in data.items():
    # Get the list of column names for the current DataFrame
    columns = df.columns
    
    # Get the list of columns to delete
    columns_to_delete = get_columns_to_delete(columns)
    
    # Create a new DataFrame without the columns to delete
    df_dropped = df.drop(columns=columns_to_delete)
    
    # Add the new DataFrame to the new dictionary
    data_dropped[key] = df_dropped

In [15]:
# Get the DataFrame with the key '16-07.csv'
df_16_07 = data_dropped['16-07.csv']

# Get the shape of the DataFrame
shape_16_07 = df_16_07.shape

# Print the shape
print("Shape of the DataFrame with key '16-07.csv':", shape_16_07)

# Print column names
print(df_16_07.columns)

Shape of the DataFrame with key '16-07.csv': (2750, 39)
Index(['frame', ' face_id', ' confidence', ' success', ' AU01_r', ' AU02_r',
       ' AU04_r', ' AU05_r', ' AU06_r', ' AU07_r', ' AU09_r', ' AU10_r',
       ' AU12_r', ' AU14_r', ' AU15_r', ' AU17_r', ' AU20_r', ' AU23_r',
       ' AU25_r', ' AU26_r', ' AU45_r', ' AU01_c', ' AU02_c', ' AU04_c',
       ' AU05_c', ' AU06_c', ' AU07_c', ' AU09_c', ' AU10_c', ' AU12_c',
       ' AU14_c', ' AU15_c', ' AU17_c', ' AU20_c', ' AU23_c', ' AU25_c',
       ' AU26_c', ' AU28_c', ' AU45_c'],
      dtype='object')


## Adding ID 

In [19]:
# Iterate over the keys of the dictionary
for key in data_dropped.keys():
    # Extract the ID from the key
    id = re.search(r'\d+', key).group()
    
    # Add a new column to the dataframe with the extracted ID
    data_dropped[key]['ID'] = id


In [27]:
# Get the DataFrame from the dictionary
df_16_07 = data_dropped['16-07.csv']

display(df_16_07)

frame   face_id   confidence   success   AU01_r   AU02_r   AU04_r  \
0         1         0         0.98         1     0.00     0.00     0.92   
1         2         0         0.98         1     0.07     0.00     0.78   
2         3         0         0.98         1     0.15     0.00     0.81   
3         4         0         0.98         1     0.14     0.10     0.82   
4         5         0         0.98         1     0.15     0.11     0.87   
...     ...       ...          ...       ...      ...      ...      ...   
2745   2746         0         0.98         1     0.17     0.10     0.54   
2746   2747         0         0.98         1     0.22     0.13     0.57   
2747   2748         0         0.98         1     0.14     0.03     0.61   
2748   2749         0         0.98         1     0.09     0.03     0.60   
2749   2750         0         0.98         1     0.10     0.01     0.68   

       AU05_r   AU06_r   AU07_r  ...   AU14_c   AU15_c   AU17_c   AU20_c  \
0         0.0     0.23     1.15  ...      1.0      0.0      0.0      0.0   
1         0.0     0.16     0.95  ...      1.0      0.0      0.0      0.0   
2         0.0     0.16     0.91  ...      1.0      0.0      1.0      0.0   
3         0.0     0.15     1.01  ...      1.0      0.0      0.0      0.0   
4         0.0     0.15     0.99  ...      1.0      0.0      0.0      0.0   
...       ...      ...      ...  ...      ...      ...      ...      ...   
2745      0.0     0.01     1.33  ...      1.0      0.0      0.0      0.0   
2746      0.0     0.01     1.29  ...      1.0      0.0      0.0      0.0   
2747      0.0     0.00     1.24  ...      1.0      0.0      0.0      0.0   
2748      0.0     0.00     1.06  ...      1.0      0.0      0.0      0.0   
2749      0.0     0.00     0.95  ...      1.0      0.0      0.0      0.0   

       AU23_c   AU25_c   AU26_c   AU28_c   AU45_c  ID  
0         1.0      0.0      0.0      0.0      0.0  16  
1         1.0      0.0      0.0      0.0      0.0  16  
2         1.0      0.0      0.0      0.0      0.0  16  
3         1.0      0.0      0.0      0.0      0.0  16  
4         1.0      0.0      0.0      0.0      0.0  16  
...       ...      ...      ...      ...      ...  ..  
2745      1.0      0.0      0.0      0.0      0.0  16  
2746      1.0      0.0      0.0      0.0      0.0  16  
2747      1.0      0.0      0.0      0.0      0.0  16  
2748      1.0      0.0      0.0      0.0      0.0  16  
2749      1.0      0.0      0.0      0.0      0.0  16  

[2750 rows x 40 columns]

## Adding Stage (timepoint)

In [21]:
def extract_timeframe(filename):
    # Extract numerical values from the filename
    numerical_values = re.findall(r'\d+', filename)
    
    # Convert each numerical value to an integer
    timeframes = [int(value) for value in numerical_values]
    
    return timeframes

In [22]:
# Test cases
filename1 = '16-07.csv'
filename2 = '6-03,04,05,06.csv'

timeframes1 = extract_timeframe(filename1)
timeframes2 = extract_timeframe(filename2)

print("Timeframes from filename 1:", timeframes1)
print("Timeframes from filename 2:", timeframes2)

Timeframes from filename 1: [16, 7]
Timeframes from filename 2: [6, 3, 4, 5, 6]


In [23]:
# Check to see if all the keys have the same format 
def check_key_format(keys):
    # Regular expression patterns to match the two formats
    pattern_1 = re.compile(r'^\d+_\d+(\,\d+)*\.csv$')
    pattern_2 = re.compile(r'^\d+(-\d+(\,\d+)*)*\.csv$')

    # Lists to store keys with different formats
    format_1_keys = []
    format_2_keys = []
    other_keys = []

    # Iterate through all keys
    for key in keys:
        # Match the key format with both patterns
        match_1 = pattern_1.match(key)
        match_2 = pattern_2.match(key)
        if match_1:
            format_1_keys.append(key)
        elif match_2:
            format_2_keys.append(key)
        else:
            other_keys.append(key)

    return format_1_keys, format_2_keys, other_keys

# Usage example:
keys = data_dropped.keys()  
format_1_keys, format_2_keys, other_keys = check_key_format(keys)
print("Keys with format 1:", format_1_keys)
print("Keys with format 2:", format_2_keys)
print("Other keys:", other_keys)


Keys with format 1: ['80_04,05,06.csv', '324_02.csv', '87_04,05,06.csv', '78_04,05,06.csv', '328_02.csv', '92_04,05,06.csv', '38_4,5,6.csv', '129_04,05,06.csv', '85_07.csv', '95_04,05,06.csv', '300_02.csv', '97_07.csv', '290_01.csv', '312_02.csv', '127_03.csv', '40_07.csv', '101_07.csv', '118_04,05,06.csv', '324_01.csv', '144_03.csv', '113_07.csv', '52_07.csv', '328_01.csv', '31_07.csv', '125_07.csv', '64_07.csv', '146_07.csv', '300_01.csv', '49_04,05,06.csv', '89_04.csv', '76_07.csv', '68_07.csv', '129_07.csv', '290_02.csv', '312_01.csv', '68_03.csv', '33_07.csv', '94_04,05,06.csv', '111_07.csv', '39_04,05,06.csv', '93_04,05,06.csv', '50_07.csv', '326_01.csv', '64_03.csv', '42_07.csv', '103_07.csv', '79_04,05,06.csv', '292_02.csv', '310_01.csv', '31_03.csv', '81_04,05,06.csv', '52_03.csv', '113_03.csv', '74_07.csv', '135_07.csv', '144_07.csv', '302_01.csv', '139_07.csv', '78_07.csv', '127_07.csv', '66_07.csv', '48_04,05,06.csv', '326_02.csv', '95_07.csv', '292_01.csv', '310_02.csv', '

Keys with a different format: Other keys: ['105_04 donation not completed. No blood flow.csv', 'DSCN2370.csv', '126_01 (5).csv', '126_01 (4).csv', '126_01 (3).csv']


In [24]:
# Filter out dataframes with keys in other_keys
data_filtered = {key: value for key, value in data_dropped.items() if key not in other_keys}

# Verify that dataframes with keys in other_keys are deleted
print("Remaining keys after deletion:", data_filtered.keys())
print(len(data_filtered))

Remaining keys after deletion: dict_keys(['16-07.csv', '80_04,05,06.csv', '7-04-05-06.csv', '324_02.csv', '87_04,05,06.csv', '78_04,05,06.csv', '328_02.csv', '92_04,05,06.csv', '38_4,5,6.csv', '129_04,05,06.csv', '85_07.csv', '95_04,05,06.csv', '300_02.csv', '20-07.csv', '97_07.csv', '290_01.csv', '312_02.csv', '127_03.csv', '40_07.csv', '101_07.csv', '118_04,05,06.csv', '14-03,04,05,06.csv', '324_01.csv', '144_03.csv', '113_07.csv', '52_07.csv', '328_01.csv', '31_07.csv', '125_07.csv', '64_07.csv', '146_07.csv', '300_01.csv', '49_04,05,06.csv', '89_04.csv', '76_07.csv', '68_07.csv', '129_07.csv', '290_02.csv', '312_01.csv', '68_03.csv', '33_07.csv', '94_04,05,06.csv', '111_07.csv', '39_04,05,06.csv', '93_04,05,06.csv', '50_07.csv', '326_01.csv', '64_03.csv', '42_07.csv', '103_07.csv', '79_04,05,06.csv', '292_02.csv', '310_01.csv', '31_03.csv', '81_04,05,06.csv', '52_03.csv', '113_03.csv', '74_07.csv', '135_07.csv', '144_07.csv', '13-03,04,05,06.csv', '302_01.csv', '139_07.csv', '78_07

We used to have 403 files (all were put into a dictionary). Then 5 dataframes in the dictionary were dropped, so we have a length of 398. 

In [28]:
# Usage example:
keys = data_filtered.keys()  
format_1_keys, format_2_keys, other_keys = check_key_format(keys)
print("Keys with format 1:", format_1_keys)
print("Keys with format 2:", format_2_keys)
print("Other keys:", other_keys)

Keys with format 1: ['80_04,05,06.csv', '324_02.csv', '87_04,05,06.csv', '78_04,05,06.csv', '328_02.csv', '92_04,05,06.csv', '38_4,5,6.csv', '129_04,05,06.csv', '85_07.csv', '95_04,05,06.csv', '300_02.csv', '97_07.csv', '290_01.csv', '312_02.csv', '127_03.csv', '40_07.csv', '101_07.csv', '118_04,05,06.csv', '324_01.csv', '144_03.csv', '113_07.csv', '52_07.csv', '328_01.csv', '31_07.csv', '125_07.csv', '64_07.csv', '146_07.csv', '300_01.csv', '49_04,05,06.csv', '89_04.csv', '76_07.csv', '68_07.csv', '129_07.csv', '290_02.csv', '312_01.csv', '68_03.csv', '33_07.csv', '94_04,05,06.csv', '111_07.csv', '39_04,05,06.csv', '93_04,05,06.csv', '50_07.csv', '326_01.csv', '64_03.csv', '42_07.csv', '103_07.csv', '79_04,05,06.csv', '292_02.csv', '310_01.csv', '31_03.csv', '81_04,05,06.csv', '52_03.csv', '113_03.csv', '74_07.csv', '135_07.csv', '144_07.csv', '302_01.csv', '139_07.csv', '78_07.csv', '127_07.csv', '66_07.csv', '48_04,05,06.csv', '326_02.csv', '95_07.csv', '292_01.csv', '310_02.csv', '

In [29]:
# Iterate through the keys and modify keys to the desired format
for key in list(data_filtered.keys()):
    parts = key.split('-')
    if len(parts) >= 2 and parts[-1].endswith('.csv'):
        id_part = parts[0]
        timepoint_part = ",".join(parts[1].split(','))
        new_key = f"{id_part}_{timepoint_part}.csv"
        data_filtered[new_key] = data_filtered.pop(key)

# Verify the modified keys
print("Modified keys in desired format:", data_filtered.keys())
print(len(data_filtered))

Modified keys in desired format: dict_keys(['80_04,05,06.csv', '324_02.csv', '87_04,05,06.csv', '78_04,05,06.csv', '328_02.csv', '92_04,05,06.csv', '38_4,5,6.csv', '129_04,05,06.csv', '85_07.csv', '95_04,05,06.csv', '300_02.csv', '97_07.csv', '290_01.csv', '312_02.csv', '127_03.csv', '40_07.csv', '101_07.csv', '118_04,05,06.csv', '324_01.csv', '144_03.csv', '113_07.csv', '52_07.csv', '328_01.csv', '31_07.csv', '125_07.csv', '64_07.csv', '146_07.csv', '300_01.csv', '49_04,05,06.csv', '89_04.csv', '76_07.csv', '68_07.csv', '129_07.csv', '290_02.csv', '312_01.csv', '68_03.csv', '33_07.csv', '94_04,05,06.csv', '111_07.csv', '39_04,05,06.csv', '93_04,05,06.csv', '50_07.csv', '326_01.csv', '64_03.csv', '42_07.csv', '103_07.csv', '79_04,05,06.csv', '292_02.csv', '310_01.csv', '31_03.csv', '81_04,05,06.csv', '52_03.csv', '113_03.csv', '74_07.csv', '135_07.csv', '144_07.csv', '302_01.csv', '139_07.csv', '78_07.csv', '127_07.csv', '66_07.csv', '48_04,05,06.csv', '326_02.csv', '95_07.csv', '292_0

Create a column which includes the stages as numerical values.

In [25]:
# Iterate through the dataframes in the data_filtered dictionary
for key, df in data_filtered.items():
    # Drop the 'timeframe' and 'Timeframe' columns if they exist
    if 'timeframe' in df.columns:
        df.drop(columns=['timeframe'], inplace=True)
    if 'Timeframe' in df.columns:
        df.drop(columns=['Timeframe'], inplace=True)
    if 'Filename' in df.columns:
        df.drop(columns=['Filename'], inplace=True)

In [30]:
# Iterate through the dataframes in the data_filtered dictionary
for key, df in data_filtered.items():
    # Remove the '.csv' suffix from the key
    key = key.replace('.csv', '')
    # Extract the timepoints from the key
    timepoints_str = key.split('_')[1]
    # Remove leading zeros from the timepoints string
    timepoints_str = ','.join(str(int(tp.lstrip('0'))) for tp in timepoints_str.split(','))
    # Convert the timepoints string into a list of integers
    timepoints_list = [int(tp) for tp in timepoints_str.split(',')]
    # Add the Timeframe column to the dataframe
    df['Timeframe'] = [timepoints_list] * len(df)

In [32]:
# Get the DataFrame from the dictionary
df_95 = data_filtered['95_04,05,06.csv']

display(df_95)

frame   face_id   confidence   success   AU01_r   AU02_r   AU04_r  \
0          1       0.0         0.98       1.0     0.00     0.00     0.00   
1          2       0.0         0.98       1.0     0.00     0.00     0.00   
2          3       0.0         0.98       1.0     0.00     0.00     0.00   
3          4       0.0         0.98       1.0     0.00     0.00     0.00   
4          5       0.0         0.98       1.0     0.00     0.00     0.00   
...      ...       ...          ...       ...      ...      ...      ...   
17880  17881       0.0         0.98       1.0     2.59     0.59     0.58   
17881  17882       0.0         0.98       1.0     2.73     0.81     0.65   
17882  17883       0.0         0.98       1.0     2.73     0.79     0.55   
17883  17884       0.0         0.98       1.0     2.77     0.82     0.44   
17884      2       NaN          NaN       NaN      NaN      NaN      NaN   

        AU05_r   AU06_r   AU07_r  ...   AU15_c   AU17_c   AU20_c   AU23_c  \
0          0.0     0.43     0.00  ...      1.0      0.0      1.0      1.0   
1          0.0     0.41     0.00  ...      1.0      0.0      1.0      1.0   
2          0.0     0.37     0.00  ...      1.0      0.0      1.0      1.0   
3          0.0     0.29     0.00  ...      1.0      0.0      1.0      1.0   
4          0.0     0.18     0.00  ...      1.0      0.0      1.0      1.0   
...        ...      ...      ...  ...      ...      ...      ...      ...   
17880      0.0     0.00     0.16  ...      0.0      0.0      0.0      0.0   
17881      0.0     0.00     0.26  ...      0.0      0.0      0.0      0.0   
17882      0.0     0.00     0.19  ...      0.0      0.0      0.0      0.0   
17883      0.0     0.00     0.32  ...      0.0      0.0      0.0      0.0   
17884      NaN      NaN      NaN  ...      NaN      NaN      NaN      NaN   

        AU25_c   AU26_c   AU28_c   AU45_c  ID  Timeframe  
0          0.0      0.0      1.0      0.0  95  [4, 5, 6]  
1          0.0      0.0      1.0      0.0  95  [4, 5, 6]  
2          0.0      0.0      1.0      0.0  95  [4, 5, 6]  
3          0.0      0.0      1.0      0.0  95  [4, 5, 6]  
4          0.0      0.0      1.0      0.0  95  [4, 5, 6]  
...        ...      ...      ...      ...  ..        ...  
17880      1.0      1.0      0.0      1.0  95  [4, 5, 6]  
17881      1.0      1.0      0.0      1.0  95  [4, 5, 6]  
17882      1.0      1.0      0.0      1.0  95  [4, 5, 6]  
17883      1.0      1.0      0.0      1.0  95  [4, 5, 6]  
17884      NaN      NaN      NaN      NaN  95  [4, 5, 6]  

[17885 rows x 41 columns]

## Merge into one big dataframe 

In [33]:
df = pd.concat(data_filtered.values(), ignore_index=True)

In [44]:
# Rename columns using a dictionary where keys are the current column names and values are the new column names
df = df.rename(columns={'frame': 'Frame',
                        ' face_id': 'Face_id', 
                        ' timestamp': 'Timestamp',
                        ' confidence': 'Confidence', 
                        ' success': 'Success'
                        # Add more columns as needed
                        })


In [45]:
display(df.head(10))

Frame  Face_id  Confidence  Success   AU01_r   AU02_r   AU04_r   AU05_r  \
103   104      0.0        0.88      1.0     0.00     0.00      0.0      0.0   
104   105      0.0        0.98      1.0     0.00     0.00      0.0      0.0   
111   112      0.0        0.98      1.0     0.31     0.59      0.0      0.0   
114   115      0.0        0.98      1.0     0.00     0.12      0.0      0.0   
153   154      0.0        0.98      1.0     1.26     1.67      0.0      0.0   
154   155      0.0        0.98      1.0     1.50     2.48      0.0      0.0   
155   156      0.0        0.88      1.0     0.88     2.77      0.0      0.0   
156   157      0.0        0.98      1.0     0.34     2.22      0.0      0.0   
159   160      0.0        0.98      1.0     0.00     1.30      0.0      0.0   
160   161      0.0        0.98      1.0     0.00     1.79      0.0      0.0   

      AU06_r   AU07_r  ...   AU15_c   AU17_c   AU20_c   AU23_c   AU25_c  \
103     1.34     0.22  ...      0.0      0.0      0.0      0.0      0.0   
104     1.34     0.22  ...      0.0      0.0      0.0      0.0      0.0   
111     0.31     0.31  ...      0.0      0.0      0.0      0.0      0.0   
114     0.32     0.06  ...      0.0      0.0      0.0      0.0      0.0   
153     0.39     0.26  ...      1.0      0.0      0.0      0.0      1.0   
154     0.60     0.26  ...      1.0      0.0      0.0      0.0      1.0   
155     0.59     0.02  ...      1.0      0.0      0.0      0.0      1.0   
156     0.73     0.00  ...      1.0      0.0      0.0      0.0      1.0   
159     0.19     0.00  ...      1.0      1.0      0.0      0.0      0.0   
160     0.25     0.00  ...      1.0      1.0      0.0      0.0      0.0   

      AU26_c   AU28_c   AU45_c  ID  Timeframe  
103      0.0      0.0      0.0  80  [4, 5, 6]  
104      0.0      0.0      0.0  80  [4, 5, 6]  
111      0.0      0.0      0.0  80  [4, 5, 6]  
114      0.0      0.0      0.0  80  [4, 5, 6]  
153      1.0      1.0      0.0  80  [4, 5, 6]  
154      1.0      1.0      0.0  80  [4, 5, 6]  
155      1.0      1.0      0.0  80  [4, 5, 6]  
156      1.0      1.0      0.0  80  [4, 5, 6]  
159      0.0      1.0      0.0  80  [4, 5, 6]  
160      0.0      1.0      0.0  80  [4, 5, 6]  

[10 rows x 41 columns]

## Drop columns with confidence below 0.80

In [47]:
print(df.shape)

(3174403, 41)


In [46]:
df = df[df['Confidence'] >= 0.80]

In [48]:
print(df.shape)

(3174403, 41)


## Check for missing values

In [49]:
# Check for missing values in 'ID' and 'Timeframe' columns
missing_id = df['ID'].isna().any()
missing_timeframe = df['Timeframe'].isna().any()

# Print the results
print("Missing values in 'ID' column:", missing_id)
print("Missing values in 'Timeframe' column:", missing_timeframe)


Missing values in 'ID' column: False
Missing values in 'Timeframe' column: False


## Check column names

In [50]:
# Get all column names in the DataFrame
column_names = df.columns.tolist()

# Print the column names
print("Column names:", column_names)
print(len(column_names))


Column names: ['Frame', 'Face_id', 'Confidence', 'Success', ' AU01_r', ' AU02_r', ' AU04_r', ' AU05_r', ' AU06_r', ' AU07_r', ' AU09_r', ' AU10_r', ' AU12_r', ' AU14_r', ' AU15_r', ' AU17_r', ' AU20_r', ' AU23_r', ' AU25_r', ' AU26_r', ' AU45_r', ' AU01_c', ' AU02_c', ' AU04_c', ' AU05_c', ' AU06_c', ' AU07_c', ' AU09_c', ' AU10_c', ' AU12_c', ' AU14_c', ' AU15_c', ' AU17_c', ' AU20_c', ' AU23_c', ' AU25_c', ' AU26_c', ' AU28_c', ' AU45_c', 'ID', 'Timeframe']
41


## Check number of unique IDs

In [51]:
# Get number of unique IDs
num_unique_ids = len(df['ID'].unique())

print("Number of Unique IDs:", num_unique_ids)

Number of Unique IDs: 184


## Saving my dataframe as a pickle file and a csv 

In [52]:
# Save big_dataframe as a pickle file
df.to_pickle("/Users/dionnespaltman/Desktop/V3/action_units_temp.pkl")

# Save big_dataframe as a CSV file
df.to_csv("/Users/dionnespaltman/Desktop/V3/action_units_temp.csv", index=False)

## Loading pickle file

In [53]:
file_path = '/Users/dionnespaltman/Desktop/downloading_new/big_dataframe.pkl'

# Read the pickle file into a DataFrame
df = pd.read_pickle(file_path)

# Display the DataFrame
display(df)

frame  face_id  confidence   success   AU01_r   AU02_r   AU04_r  \
0         104      0.0        0.88       1.0     0.00     0.00     0.00   
1         105      0.0        0.98       1.0     0.00     0.00     0.00   
2         112      0.0        0.98       1.0     0.31     0.59     0.00   
3         115      0.0        0.98       1.0     0.00     0.12     0.00   
4         154      0.0        0.98       1.0     1.26     1.67     0.00   
...       ...      ...         ...       ...      ...      ...      ...   
3395911  2196      0.0        0.98       1.0     0.62     0.36     0.30   
3395912  2197      0.0        0.98       1.0     0.63     0.38     0.25   
3395913  2198      0.0        0.98       1.0     0.69     0.51     0.26   
3395914  2199      0.0        0.98       1.0     0.67     0.46     0.32   
3395915  2200      0.0        0.98       1.0     0.62     0.45     0.45   

          AU05_r   AU06_r   AU07_r  ...   AU15_c   AU17_c   AU20_c   AU23_c  \
0            0.0     1.34     0.22  ...      0.0      0.0      0.0      0.0   
1            0.0     1.34     0.22  ...      0.0      0.0      0.0      0.0   
2            0.0     0.31     0.31  ...      0.0      0.0      0.0      0.0   
3            0.0     0.32     0.06  ...      0.0      0.0      0.0      0.0   
4            0.0     0.39     0.26  ...      1.0      0.0      0.0      0.0   
...          ...      ...      ...  ...      ...      ...      ...      ...   
3395911      0.0     0.70     0.47  ...      0.0      1.0      0.0      1.0   
3395912      0.0     0.78     0.56  ...      0.0      1.0      0.0      1.0   
3395913      0.0     0.79     0.56  ...      0.0      1.0      0.0      1.0   
3395914      0.0     0.81     0.56  ...      1.0      1.0      0.0      1.0   
3395915      0.0     0.81     0.57  ...      0.0      1.0      0.0      1.0   

          AU25_c   AU26_c   AU28_c   AU45_c  ID  Timeframe  
0            0.0      0.0      0.0      0.0  80  [4, 5, 6]  
1            0.0      0.0      0.0      0.0  80  [4, 5, 6]  
2            0.0      0.0      0.0      0.0  80  [4, 5, 6]  
3            0.0      0.0      0.0      0.0  80  [4, 5, 6]  
4            1.0      1.0      1.0      0.0  80  [4, 5, 6]  
...          ...      ...      ...      ...  ..        ...  
3395911      0.0      0.0      0.0      0.0  25        [7]  
3395912      0.0      0.0      0.0      0.0  25        [7]  
3395913      0.0      0.0      0.0      0.0  25        [7]  
3395914      0.0      0.0      0.0      0.0  25        [7]  
3395915      0.0      0.0      0.0      0.0  25        [7]  

[3395916 rows x 41 columns]